### PR TITLE
Implement clear(), erase() and __delitem__ for Dataset

### LINKED ISSUE #389: Dataset erase
Implement `erase`.

See https://github.com/scipp/scipp/tree/v0.0.1 and in particular dataset.h (before the refactor) for what was previously supported.

### PR DESCRIPTION
**Pull request by DanNixon**
Fixes #389

### CODE REVIEW

File: python/dataset.cpp [hunk 1]
```diff
@@ -116,6 +116,7 @@ void init_dataset(py::module &m) {
       .def("__setitem__",
            [](Dataset &self, const std::string &name,
               const DataConstProxy &data) { self.setData(name, data); })
+      .def("__delitem__", &Dataset::erase)
```
Comment (SimonHeybrock) [inline python/dataset.cpp]: This should throw, to be consistent with a Python `dict`. I think this consistency is more important that the consistency with `std::map`, so I'd go with the simple/brief option of throwing in `erase`.

Comment (DanNixon) [reply python/dataset.cpp]: :+1:

File: core/include/scipp/core/dataset.h [hunk 2]
```diff
@@ -247,6 +249,8 @@ class SCIPP_CORE_EXPORT Dataset {
 
   bool contains(const std::string_view name) const noexcept;
 
+  size_t erase(const std::string_view name);
```
[the remark this answers is not in the document]
Comment (DanNixon) [reply core/include/scipp/core/dataset.h]: :+1:

File: core/dataset.cpp [hunk 3]
```diff
@@ -196,6 +215,26 @@ void Dataset::setDims(const Dimensions &dims, const Dim coordDim) {
...
 }
 
+void Dataset::rebuildDims() {
+  /* Clear old extents */
+  m_dims.clear();
+
+  for (const auto &d : m_data) {
+    if (d.second.data) {
+      setDims(d.second.data->dims());
+    }
```
[the remark this answers is not in the document]
Comment (DanNixon) [reply core/dataset.cpp]: :+1:

**Comment by SimonHeybrock** [comment]
Actually, shouldn't we also have `clear()` available in Python? I think it is missing here?

**Comment by SimonHeybrock** [comment]
Unfortunately there is a merge conflict now. :(

**Pull request merged**